doc/poddoc/man/btmdisp.8
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "BTMDISP 8"
.TH BTMDISP 8 "2009-05-18" "GNUbatch Release 1" "GNUbatch Batch Scheduler"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
btmdisp \- generate messages
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBbtmdisp\fR
\&\fIoptions\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBBtmdisp\fR generates messages as required by \fIbtsched\fR\|(8) in response to
the mail or write completion options of batch jobs.
.PP
By default, it uses the system basic mailer to dispose of mail
options, \fIbtwrite\fR\|(8) to send messages to users' terminals and
\&\fIdosbtwrite\fR\|(8) to send messages to Windows PCs.
.PP
The messages are generated by \fBbtmdisp\fR from the system message file,
which by default is \fIbtint-config\fR.
.PP
The program to be used in each case may be overridden by assignments
to the environment variables \f(CW\*(C`MAILER\*(C'\fR, \f(CW\*(C`WRITER\*(C'\fR and \f(CW\*(C`DOSWRITER\*(C'\fR,
most conveniently in the master configuration file
\&\fIgnubatch.conf\fR. The program (or shell script) to be run in each
case should take data on standard input and the relevant user name as
the first argument, and will run under the identity \f(CW\*(C`batch\*(C'\fR.
.PP
These variables may also be set on a per-user basis by assignment in a
\&\f(CW\*(C`.gnubatch\*(C'\fR file located in a user's home directory. The user may also
specify an alternative message file by assignment to the variable
\&\f(CW\*(C`SYSMESG\*(C'\fR. These programs or scripts will be run under the identity
of the user, typically the owner of the job to be run.
.PP
The interface (options etc) are internal to \fBGNUbatch\fR and are not
documented here.
.SH "FILES"
.IX Header "FILES"
\&\fIbtint-config\fR
message file
.PP
\&\fI~/.gnubatch\fR
user's home directory
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.IP "\s-1DOSWRITER\s0" 4
.IX Item "DOSWRITER"
alternative \fIwrite to \s-1PC\s0\fR program or script.
.IP "\s-1MAILER\s0" 4
.IX Item "MAILER"
alternative \fImail results\fR program or script.
.IP "\s-1WRITER\s0" 4
.IX Item "WRITER"
alternative \fIwrite results\fR program or script.
.IP "\s-1SPHELPDIR\s0" 4
.IX Item "SPHELPDIR"
alternative location for help file directory, by default
\&\fI/usr/spool/progs\fR.
.IP "\s-1SYSMESG\s0" 4
.IX Item "SYSMESG"
alternative message file location (relative to user's home directory
if need be).
.SH "NOTES"
.IX Header "NOTES"
\&\fBBtmdisp\fR is identical to \fBGNUspool\fR's \fIspmdisp\fR\|(8), apart from using
different message and configuration files.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIdosbtwrite\fR\|(8),
\&\fIbtsched\fR\|(8),
\&\fIbtwrite\fR\|(8).
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2009 Free Software Foundation, Inc.
This is free software. You may redistribute copies of it under the
terms of the \s-1GNU\s0 General Public License
<http://www.gnu.org/licenses/gpl.html>.
There is \s-1NO\s0 \s-1WARRANTY\s0, to the extent permitted by law.
.SH "AUTHOR"
.IX Header "AUTHOR"
John M Collins, Xi Software Ltd.


debian/changelog
gnubatch (1.5) unstable; urgency=low

  * Release 1.5
    Revise network handling, increase user/host lengths in API.
    Mostly rewrite xbnetserv and half of sh_network.c
    Fix bug with expansion of names in redirections.

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Tue, 26 Jun 2012 12:45:06 +0100

gnubatch (1.1) unstable; urgency=low

  * Initial release (Closes: #572633)

 -- John M Collins (Xi Software Ltd) <jmc@xisl.com>  Fri, 05 Mar 2010 11:22:06 +0000



src/btmdisp.c
/* btmdisp.c -- message dispatcher

   Copyright 2009 Free Software Foundation, Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include "config.h"
#include <stdio.h>
#include "incl_sig.h"
#include <sys/types.h>
#ifdef	HAVE_FCNTL_H
#include <fcntl.h>
#endif
#include <errno.h>
#ifdef	TIME_WITH_SYS_TIME
#include <sys/time.h>
#include <time.h>
#elif	defined(HAVE_SYS_TIME_H)
#include <sys/time.h>
#else
#include <time.h>
#endif
#include "defaults.h"
#include "errnums.h"
#include "files.h"
#include "ecodes.h"
#include "incl_unix.h"
#include "incl_ugid.h"
#include "cfile.h"
#include "notify.h"

static	char	Filename[] = __FILE__;

char	configname[] = USER_CONFIG;
char	shellname[] = DEF_CI_PATH;

char	*writer, *doswriter, *mailer;
char	*jobtitle, *exec_host;
char	*batch_uname, *homedir;
jobno_t	jobnum;

int_ugid_t	spuid, batch_uid;

int	has_sofile = 0, has_sefile = 0, pass_thru = 0;
int	repl_mail = 0, repl_write = 0, repl_doswrite = 0;
int	msg_code = $PE{Job completed msg}, exit_stat;
char	*orig_host;

extern	char	*Helpfile_path;

void  nomem(const char *fl, const int ln)
{
	fprintf(stderr, "%s:Mem alloc fault: %s line %d\n", progname, fl, ln);
	exit(E_NOMEM);
}

void  pushfile(FILE *ofl, const int ifd, const int mcode)
{
	FILE	*ifl;
	int	ch;

	if  ((ifl = fdopen(ifd, "r")))  {
		if  ((ch = getc(ifl)) != EOF)  {
			fprint_error(ofl, mcode);
			do	putc(ch, ofl);
			while  ((ch = getc(ifl)) != EOF);
		}
		fclose(ifl);
	}
}

/* Invoke mail or write command.  Don't use popen since there is a bug
   in it which makes funny things happen if file descriptors 0 or
   1 are closed.  */

void  rmsg(char *cmd)
{
	int	ncode;
	FILE	*po;
	int	pfds[2];
	PIDTYPE	pid;

	if  (pipe(pfds) < 0  || (pid = fork()) < 0)
		return;

	if  (pid == 0)  {		/*  Child process  */
		char	*cp, **ap, *arglist[5];

		close(pfds[1]);	/*  Write side  */
		if  (pfds[0] != 0)  {
			close(0);
			Ignored_error = dup(pfds[0]);
			close(pfds[0]);
		}

		ap = arglist;
		if  ((cp = strrchr(cmd, '/')))
			cp++;
		else
			cp = cmd;
		*ap++ = cp;
		if  (orig_host)
			*ap++ = orig_host;
		*ap++ = batch_uname;
		*ap = (char *) 0;

		execv(cmd, arglist);
		if  (errno == ENOEXEC)  {
			ap = arglist;
			*ap++ = cp;
			*ap++ = cmd;
			if  (orig_host)
				*ap++ = orig_host;
			*ap++ = batch_uname;
			*ap = (char *) 0;
			execv(shellname, arglist);
		}
		exit(255);
	}

	close(pfds[0]);			/*  Read side  */
	if  ((po = fdopen(pfds[1], "w")) == (FILE *) 0)  {
		kill(pid, SIGKILL);
		return;
	}

	if  (pass_thru)  {
		int  ch;
		while  ((ch = getchar()) != EOF)
			putc(ch, po);
	}
	else  {
		disp_arg[0] = jobnum;
		disp_arg[1] = exit_stat & 127;
		disp_arg[2] = (exit_stat >> 8) & 255;
		disp_str = exec_host;
		disp_str2 = jobtitle;
		ncode = msg_code;
		if  (jobtitle[0] == '\0')
			ncode -= $S{No title offset};
		if  (exec_host[0])
			ncode += $S{Host name offset};
		fprint_error(po, ncode);
		if  (has_sofile)
			pushfile(po, SOSTREAM, $E{Job stdout was});
		if  (has_sefile)
			pushfile(po, SESTREAM, $E{Job stderr was});
	}
	fclose(po);
	exit(0);
}

FILE *getmsgfile()
{
	FILE		*res;
	char		*homedf, *sysf, *repl;
	unsigned	hdlng, lng;

	/* If I don't know the user, just return the standard file.  */

	if  ((spuid = lookup_uname(batch_uname)) == UNKNOWN_UID)  {
		homedir = "/";
		return  open_icfile();
	}

	/* Get user's home directory */

	homedir = unameproc("~", "/", (uid_t) spuid);
	hdlng = strlen(homedir) + 1;
	lng = hdlng + sizeof(configname);

	/* Get hold of .xibatch file in home directory for user.  */

	if  (!(homedf = malloc(lng)))
		ABORT_NOMEM;
	strcpy(homedf, homedir);
	strcat(homedf, "/");
	strcat(homedf, configname);

	/* Whilst we're there, grab alternate programs if
	   specified. Also reset uid to the relevant user.  */

	if  ((repl = rdoptfile(homedf, "MAILER")))  {
		mailer = repl;
		repl_mail++;
	}
	if  ((repl = rdoptfile(homedf, "WRITER")))  {
		writer = repl;
		repl_write++;
	}
	if  ((repl = rdoptfile(homedf, "DOSWRITER")))  {
		doswriter = repl;
		repl_doswrite++;
	}
	if  (!(sysf = rdoptfile(homedf, "SYSMESG")))  {
		free(homedf);
		return  open_icfile();
	}

	/* If not absolute, bring it down from the home directory.  */

	if  (sysf[0] != '/')  {
		char	*abssysf;
		lng = hdlng + strlen(sysf) + 1;
		if  (!(abssysf = malloc(lng)))
			ABORT_NOMEM;
		strcpy(abssysf, homedir);
		strcat(abssysf, "/");
		strcat(abssysf, sysf);
		free(sysf);	/* Not needed any more */
		sysf = abssysf;
	}

	free(homedf);

	if  (!(res = fopen(sysf, "r")))  {
		free(sysf);
		return  open_icfile();
	}

	Helpfile_path = sysf;
	fcntl(fileno(res), F_SETFD, 1);
	return  res;
}

MAINFN_TYPE  main(int argc, char **argv)
{
	cmd_type	cmd = NOTIFY_MAIL;
	char		**ep, **lep;
	extern	char	**environ;
	static	char	lnam[] = "LOGNAME=";

	versionprint(argv, "$Revision: 1.5 $", 1);

	if  ((progname = strrchr(argv[0], '/')))
		progname++;
	else
		progname = argv[0];

	init_mcfile();

	/* Initialise default message despatch */

	writer = envprocess(WRITER);
	doswriter = envprocess(DOSWRITER);
	mailer = envprocess(MAILER);

	if  ((batch_uid = lookup_uname(BATCHUNAME)) == UNKNOWN_UID)
		batch_uid = ROOTID;

	/* Now decode arguments.
	   -m	Mail style
	   -w	Write style
	   -d	Dos write
	   -o	Standard output file sent on f.d. SOSTREAM
	   -e	Standard error file sent on f.d. SESTREAM
	   -x	Pass through (no message code)
	   Follow by user id of job.
	   Follow by message code (integer) and possible host name.
	   Any other arguments are read from environment.
	   We don't check the arguments very well. */

	while  (argv[1])  {
		char	*arg = argv[1];
		if  (*arg++ != '-')
			break;
		argv++;
		while  (*arg)
			switch  (*arg++)  {
			default:
				continue;
			case  'm':case  'M':
				cmd = NOTIFY_MAIL;
				continue;
			case  'w':case  'W':
				cmd = NOTIFY_WRITE;
				continue;
			case  'd':case  'D':
				cmd = NOTIFY_DOSWRITE;
				continue;
			case  'o':case  'O':
				has_sofile = 1;
				continue;
			case  'e':case  'E':
				has_sefile = 1;
				continue;
			case  'x':case  'X':
				pass_thru = 1;
				continue;
			}
	}

	argv++;
	if  (!*argv)
		exit(E_USAGE);
	if  (!pass_thru)  {
		jobnum = atol(*argv++);
		jobtitle = *argv++;
		exec_host = *argv++;
		if  (!jobtitle  ||  !exec_host)
			exit(E_USAGE);
	}
	batch_uname = *argv++;
	if  (!pass_thru)  {
		if  (!*argv)
			exit(E_USAGE);
		exit_stat = atoi(*argv++);
		if  (!*argv)
			exit(E_USAGE);
		msg_code = atoi(*argv++);
		if  (*argv)	/* Optional */
			orig_host = *argv++;
	}

	/* Get message file.  */

	if  (!(Cfile = getmsgfile()))
		exit(E_SETUP);

	/* And now a whole load of nonsense to delete LOGNAME from the
	   environment. For some idiotic reason mail uses this
	   rather than the uid, and as a result will pick up some
	   fossil LOGNAME which happens to be lying around.  */

	for  (ep = environ;  *ep;  ep++)  {
		if  (strncmp(*ep, lnam, sizeof(lnam) - 1) == 0)  {
			for  (lep = ep + 1;  *lep;  lep++)
				;
			*ep = *--lep;
			*lep = (char *) 0;
			break;
		}
	}

	switch  (cmd)  {
	case  NOTIFY_MAIL:
		setuid((uid_t) repl_mail? spuid: batch_uid);
		rmsg(mailer);
		break;
	case  NOTIFY_WRITE:
		setuid((uid_t) repl_write? spuid: batch_uid);
		rmsg(writer);
		break;
	case  NOTIFY_DOSWRITE:
		setuid((uid_t) repl_doswrite? spuid: batch_uid);
		rmsg(doswriter);
		break;
	}

	return  E_SETUP;		/* Actually only reached if buggy*/
}
